src/header/map.rs

```diff
@@ -14,11 +14,48 @@ use super::HeaderValue;
 pub use self::as_header_name::AsHeaderName;
 pub use self::into_header_name::IntoHeaderName;
 
-/// A set of HTTP headers
+/// The `HeaderMap` type is a specialized
+/// [multimap](<https://en.wikipedia.org/wiki/Multimap>) structure for storing
+/// header names and values.
 ///
-/// `HeaderMap` is a multimap of [`HeaderName`] to values.
+/// # Overview
+///
+/// `HeaderMap` is designed specifically for efficient manipulation of HTTP
+/// headers. It supports multiple values per header name and provides
+/// specialized APIs for insertion, retrieval, and iteration.
+///
+/// The internal implementation is optimized for common usage patterns in HTTP,
+/// and may change across versions. For example, the current implementation uses
+/// [Robin Hood
+/// hashing](<https://en.wikipedia.org/wiki/Hash_table#Robin_Hood_hashing>) to
+/// store entries compactly and enable high load factors with good performance.
+/// However, the collision resolution strategy and storage mechanism are not
+/// part of the public API and may be altered in future releases.
+///
+/// # Iteration order
+///
+/// Unless otherwise specified, the order in which items are returned by
+/// iterators from `HeaderMap` methods is arbitrary; there is no guaranteed
+/// ordering between elements yielded by such an iterator. However, for a given
+/// crate version, the iteration order will be consistent across all platforms.
+/// Changing the iteration order of such an iterator is considered as a [minor
+/// breaking change](<https://doc.rust-lang.org/cargo/reference/semver.html>).
```

Comment:
I wouldn't claim this is a minor breaking change. We strive not to make breaking changes, so claiming this would be means we couldn't change it. And since it's always been unspecified, we can continue to make that be the truth: it could change at any moment, and it cannot be depended on.